source/program/main.cpp

```diff
@@ -146,55 +146,141 @@ int multiworld_init(lua_State* L) {
         }
-        buffer[0] = output_success;        
-        buffer[1] = response & 0xff;
-        buffer[2] = (response >> 8)  & 0xff;
-        buffer[3] = (response >> 16) & 0xff;
-        return response + 4;
+        sendBuffer->insert(sendBuffer->begin(), PACKET_REMOTE_LUA_EXEC);
+        sendBuffer->insert(sendBuffer->begin() + 1, outputSuccess);
+        sendBuffer->insert(sendBuffer->begin() + 2, resultSize & 0xff);
+        sendBuffer->insert(sendBuffer->begin() + 3, (resultSize >> 8)  & 0xff);
+        sendBuffer->insert(sendBuffer->begin() + 4, (resultSize >> 16)  & 0xff);
+        return sendBuffer;
     });
 
     // Register calling update again
     multiworld_schedule_update(L);
     return 0;
 }
 
+std::vector<u8>* create_packet_from_lua_string(lua_State* L) {
+    size_t resultSize = 0;          // length of the lua string response (without \0)
+    std::vector<u8>* sendBuffer;  // sendBuffer to store the result. this pointer is returned 
+    
+    const char* luaResult = lua_tolstring(L, 1, &resultSize);
+
+    sendBuffer = new std::vector<u8>();
+    if (sendBuffer == NULL) return NULL;
+
+    uint32_t resultAs32Bit = resultSize;
+    u8* chars = reinterpret_cast<u8*>(&resultAs32Bit);
+    // If you move the following line by one line down the compiler throws a warning?!?
+    sendBuffer->insert(sendBuffer->begin(), luaResult, luaResult + resultAs32Bit);
+    sendBuffer->insert(sendBuffer->begin(), chars, chars + sizeof(uint32_t));
+
+    return sendBuffer;
+}
+
+/* This function gets called by the lua to sent message to the client */
+int gamelog_send(lua_State* L) {
+    if (RemoteApi::clientSubs.logging) {
+        RemoteApi::SendMessage([=]() -> std::vector<u8>* {
+            std::vector<u8>* sendBuffer = create_packet_from_lua_string(L); 
+            if (sendBuffer == NULL) return NULL;
+            sendBuffer->insert(sendBuffer->begin(), (u8)PACKET_LOG_MESSAGE);
+            return sendBuffer;
+        });
+    }
+    return 1;
+}
+
+/* This function gets called by the lua to sent message to the client */
+int inventory_send(lua_State* L) {
+    if (RemoteApi::clientSubs.multiWorld) {
+        RemoteApi::SendMessage([=]() -> std::vector<u8>* {
+            std::vector<u8>* sendBuffer = create_packet_from_lua_string(L); 
+            if (sendBuffer == NULL) return NULL;
+            sendBuffer->insert(sendBuffer->begin(), (u8)PACKET_NEW_INVENTORY);
+            return sendBuffer;
+        });
+    }
+    return 1;
+}
+
+/* This function gets called by the lua to sent message to the client */
+int indices_send(lua_State* L) {
+    if (RemoteApi::clientSubs.multiWorld) {
+        RemoteApi::SendMessage([=]() -> std::vector<u8>* {
+            std::vector<u8>* sendBuffer = create_packet_from_lua_string(L); 
+            if (sendBuffer == NULL) return NULL;
+            sendBuffer->insert(sendBuffer->begin(), (u8)PACKET_COLLECTED_INDICES);

```

Comment:
If these 3 functions are going to be identical, maybe this constant should be sent from lua?

Reply:
Yeah, I started implementing it and they all ended being the same function.
Anyways, I moved the duplicate code to another function and just do a minimum in these funcitons.